pkg/controllers/user/alert/configsyncer/configsyncer.go

```diff
@@ -0,0 +1,252 @@
+package configsyner
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/rancher/norman/controller"
+	alertconfig "github.com/rancher/rancher/pkg/controllers/user/alert/config"
+	"github.com/rancher/rancher/pkg/controllers/user/alert/manager"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func NewConfigSyncer(cluster *config.UserContext, manager *manager.Manager) *ConfigSyncer {
+	return &ConfigSyncer{
+		secretClient:       cluster.Core.Secrets("cattle-alerting"),
+		clusterAlertLister: cluster.Management.Management.ClusterAlerts(cluster.ClusterName).Controller().Lister(),
+		projectAlertLister: cluster.Management.Management.ProjectAlerts("").Controller().Lister(),
+		notifierLister:     cluster.Management.Management.Notifiers(cluster.ClusterName).Controller().Lister(),
+		clusterName:        cluster.ClusterName,
+		alertManager:       manager,
+	}
+
+}
+
+type ConfigSyncer struct {
+	secretClient       v1.SecretInterface
+	projectAlertLister v3.ProjectAlertLister
+	clusterAlertLister v3.ClusterAlertLister
+	notifierLister     v3.NotifierLister
+	clusterName        string
+	alertManager       *manager.Manager
+}
+
+func (d *ConfigSyncer) ProjectSync(key string, alert *v3.ProjectAlert) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) ClusterSync(key string, alert *v3.ClusterAlert) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) NotifierSync(key string, alert *v3.Notifier) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) sync() error {
+	//logrus.Info("start sync config")
+
+	clusterAlerts, err := d.clusterAlertLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "Creating cluster alerts")
+	}
+
+	projectAlerts, err := d.projectAlertLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "Creating project alerts")
+	}
+
+	pAlerts := []*v3.ProjectAlert{}
+	for _, alert := range projectAlerts {
+		if controller.ObjectInCluster(d.clusterName, alert) {
+			pAlerts = append(pAlerts, alert)
+		}
+	}
+
+	notifiers, err := d.notifierLister.List("", labels.NewSelector())
```

Comment:
do you think it would make sense to do all the list operations first? so if error occurs, we don't have to do slice operation on lines 67-73.

Reply:
Yes, it would be better to do all the list operations first., thanks for the suggestion.